File: frontend/maya-frontend-simple/src/app/services/dashboard.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';

export interface DashboardOverview {
  totalReviews: number;
  activeRepositories: number;
  averageQualityScore: number;
  criticalIssues: number;
  connectionLeaks: number;
  completionRate: number;
  lastUpdate: string;
  quickStats: QuickStat[];
}

export interface QuickStat {
  name: string;
  value: any;
  unit: string;
  trend: string;
  changePercent: number;
}

export interface QualityMetrics {
  averageScore: number;
  medianScore: number;
  scoreDistribution: number;
  totalFiles: number;
  analyzedFiles: number;
  issuesByType: { [key: string]: number };
  scoresByRepository: { [key: string]: number };
  trends: QualityTrend[];
}

export interface QualityTrend {
  date: string;
  score: number;
  issues: number;
  reviews: number;
}

export interface RepositoryRanking {
  repositoryId?: number;
  repositoryName: string;
  averageScore: number;
  totalReviews: number;
  criticalIssues: number;
  trend: number;
  status: string;
}

export interface DeveloperRanking {
  developer: string;
  averageScore: number;
  totalReviews: number;
  criticalIssues: number;
  improvement: number;
  level: string;
}

export interface ChartData {
  chartType: string;
  labels: string[];
  datasets: DataSeries[];
  options: { [key: string]: any };
}

export interface DataSeries {
  label: string;
  data: any[];
  backgroundColor: string;
  borderColor: string;
  styling: { [key: string]: any };
}

export interface SystemAlert {
  id: string;
  type: string;
  severity: string;
  message: string;
  details: string;
  timestamp: string;
  resolved: boolean;
}

@Injectable({
  providedIn: 'root'
})
export class DashboardService {
  private apiUrl = `${environment.apiUrl}/dashboard`;

  constructor(private http: HttpClient) {}

  /**
   * Obter overview geral
   */
  getOverview(days: number = 30): Observable<DashboardOverview> {
    const params = new HttpParams().set('days', days.toString());
    return this.http.get<DashboardOverview>(`${this.apiUrl}/overview`, { params });
  }

  /**
   * Obter métricas de qualidade
   */
  getQualityMetrics(days: number = 30, repositoryId?: number): Observable<QualityMetrics> {
    let params = new HttpParams().set('days', days.toString());
    if (repositoryId) {
      params = params.set('repositoryId', repositoryId.toString());
    }
    return this.http.get<QualityMetrics>(`${this.apiUrl}/quality-metrics`, { params });
  }

  /**
   * Obter análise de tendências
   */
  getTrends(days: number = 30, period: string = 'DAILY'): Observable<any> {
    const params = new HttpParams()
      .set('days', days.toString())
      .set('period', period);
    return this.http.get(`${this.apiUrl}/trends`, { params });
  }

  /**
   * Obter ranking de repositórios
   */
  getRepositoryRanking(days: number = 30, limit: number = 10): Observable<RepositoryRanking[]> {
    const params = new HttpParams()
      .set('days', days.toString())
      .set('limit', limit.toString());
    return this.http.get<RepositoryRanking[]>(`${this.apiUrl}/repository-ranking`, { params });
  }

  /**
   * Obter ranking de desenvolvedores
   */
  getDeveloperRanking(days: number = 30, limit: number = 10): Observable<DeveloperRanking[]> {
    const params = new HttpParams()
      .set('days', days.toString())
      .set('limit', limit.toString());
    return this.http.get<DeveloperRanking[]>(`${this.apiUrl}/developer-ranking`, { params });
  }

  /**
   * Obter estatísticas de problemas
   */
  getIssueStatistics(days: number = 30, repositoryId?: number): Observable<any> {
    let params = new HttpParams().set('days', days.toString());
    if (repositoryId) {
      params = params.set('repositoryId', repositoryId.toString());
    }
    return this.http.get(`${this.apiUrl}/issue-statistics`, { params });
  }

  /**
   * Obter métricas de performance
   */
  getPerformanceMetrics(days: number = 7): Observable<any> {
    const params = new HttpParams().set('days', days.toString());
    return this.http.get(`${this.apiUrl}/performance-metrics`, { params });
  }

  /**
   * Obter alertas ativos
   */
  getActiveAlerts(): Observable<SystemAlert[]> {
    return this.http.get<SystemAlert[]>(`${this.apiUrl}/alerts`);
  }

  /**
   * Obter dados para gráficos
   */
  getChartData(chartType: string, days: number = 30, repositoryId?: number): Observable<ChartData> {
    let params = new HttpParams().set('days', days.toString());
    if (repositoryId) {
      params = params.set('repositoryId', repositoryId.toString());
    }
    return this.http.get<ChartData>(`${this.apiUrl}/charts/${chartType}`, { params });
  }

  /**
   * Exportar dados do dashboard
   */
  exportDashboard(format: string, sections: string[], days: number = 30, options?: any): Observable<any> {
    const request = {
      format,
      sections,
      days,
      options: options || {}
    };
    return this.http.post(`${this.apiUrl}/export`, request);
  }
}

File: backend/src/main/java/com/sinqia/maya/service/DashboardService.java
package com.sinqia.maya.service;

import com.sinqia.maya.controller.DashboardController;
import com.sinqia.maya.entity.CodeReview;
import com.sinqia.maya.entity.DashboardMetric;
import com.sinqia.maya.repository.CodeReviewRepository;
import com.sinqia.maya.repository.RepositoryRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Serviço para dashboard e métricas
 */
@Service
@Slf4j
@RequiredArgsConstructor
public class DashboardService {

    private final CodeReviewRepository codeReviewRepository;
    private final RepositoryRepository repositoryRepository;

    /**
     * Obter overview geral do sistema
     */
    public DashboardController.DashboardOverview getOverview(int days) {
        log.debug("Gerando overview para {} dias", days);

        LocalDateTime cutoffDate = LocalDateTime.now().minusDays(days);

        long totalReviews = codeReviewRepository.countByCreatedAtAfter(cutoffDate);
        long activeRepositories = repositoryRepository.count();
        long completedReviews = codeReviewRepository.countByStatusAndCreatedAtAfter(
                CodeReview.ReviewStatus.COMPLETED, cutoffDate);
        long criticalIssues = codeReviewRepository.countCriticalIssuesAfter(cutoffDate);
        
        Double averageScore = codeReviewRepository.findAverageScoreAfter(cutoffDate);
        if (averageScore == null) averageScore = 0.0;
        
        double completionRate = totalReviews > 0 ? (double) completedReviews / totalReviews * 100 : 0.0;

        List<DashboardController.QuickStat> quickStats = generateQuickStats(cutoffDate);

        return new DashboardController.DashboardOverview(
                totalReviews,
                activeRepositories,
                averageScore,
                criticalIssues,
                0L, // TODO: Implementar contagem de vazamentos de conexão
                completionRate,
                LocalDateTime.now(),
                quickStats
        );
    }

    /**
     * Obter métricas de qualidade
     */
    public DashboardController.QualityMetrics getQualityMetrics(int days, Long repositoryId) {
        log.debug("Gerando métricas de qualidade para {} dias, repo: {}", days, repositoryId);

        LocalDateTime cutoffDate = LocalDateTime.now().minusDays(days);

        // TODO: Implementar queries específicas baseadas no repositoryId
        Double averageScore = codeReviewRepository.findAverageScoreAfter(cutoffDate);
        if (averageScore == null) averageScore = 0.0;

        Map<String, Long> issuesByType = new HashMap<>();
        issuesByType.put("CRITICAL", codeReviewRepository.countCriticalIssuesAfter(cutoffDate));
        issuesByType.put("HIGH", 0L); // TODO: Implementar
        issuesByType.put("MEDIUM", 0L); // TODO: Implementar
        issuesByType.put("LOW", 0L); // TODO: Implementar

        Map<String, Double> scoresByRepository = new HashMap<>();
        // TODO: Implementar scores por repositório

        List<DashboardController.QualityTrend> trends = new ArrayList<>();
        // TODO: Implementar tendências

        return new DashboardController.QualityMetrics(
                averageScore,
                averageScore, // TODO: Calcular mediana
                0.0, // TODO: Calcular distribuição
                0L, // TODO: Total de arquivos
                0L, // TODO: Arquivos analisados
                issuesByType,
                scoresByRepository,
                trends
        );
    }

    /**
     * Obter análise de tendências
     */
    public DashboardController.TrendAnalysis getTrends(int days, String period) {
        log.debug("Gerando tendências para {} dias, período: {}", days, period);

        // TODO: Implementar análise de tendências completa
        List<DashboardController.TrendPoint> reviewsTrend = new ArrayList<>();
        List<DashboardController.TrendPoint> qualityTrend = new ArrayList<>();
        List<DashboardController.TrendPoint> issuesTrend = new ArrayList<>();

        Map<String, Object> insights = new HashMap<>();
        insights.put("trend", "stable");
        insights.put("recommendation", "Manter qualidade atual");

        return new DashboardController.TrendAnalysis(
                period,
                reviewsTrend,
                qualityTrend,
                issuesTrend,
                insights
        );
    }

    /**
     * Obter ranking de repositórios
     */
    public List<DashboardController.RepositoryRanking> getRepositoryRanking(int days, int limit) {
        log.debug("Gerando ranking de repositórios para {} dias, limit: {}", days, limit);

        LocalDateTime cutoffDate = LocalDateTime.now().minusDays(days);
        List<Object[]> rawMetrics = codeReviewRepository.findRepositoryMetrics(cutoffDate);

        return rawMetrics.stream()
                .limit(limit)
                .map(row -> new DashboardController.RepositoryRanking(
                        null, // TODO: ID do repositório
                        (String) row[0], // nome
                        ((Number) row[2]).doubleValue(), // score médio
                        ((Number) row[1]).longValue(), // total reviews
                        ((Number) row[3]).longValue(), // issues críticos
                        0.0, // TODO: Calcular trend
                        "active"
                ))
                .toList();
    }

    /**
     * Obter ranking de desenvolvedores
     */
    public List<DashboardController.DeveloperRanking> getDeveloperRanking(int days, int limit) {
        log.debug("Gerando ranking de desenvolvedores para {} dias, limit: {}", days, limit);

        LocalDateTime cutoffDate = LocalDateTime.now().minusDays(days);
        List<Object[]> rawAuthors = codeReviewRepository.findTopAuthorsByReviewCount(cutoffDate);

        return rawAuthors.stream()
                .limit(limit)
                .map(row -> new DashboardController.DeveloperRanking(
                        (String) row[0], // autor
                        0.0, // TODO: Score médio por autor
                        ((Number) row[1]).longValue(), // total reviews
                        0L, // TODO: Issues críticos por autor
                        0.0, // TODO: Melhoria
                        "intermediate"
                ))
                .toList();
    }

    /**
     * Obter estatísticas de problemas
     */
    public DashboardController.IssueStatistics getIssueStatistics(int days, Long repositoryId) {
        log.debug("Gerando estatísticas de problemas para {} dias, repo: {}", days, repositoryId);

        Map<String, Long> issuesByType = new HashMap<>();
        issuesByType.put("CONNECTION_LEAK", 0L);
        issuesByType.put("SECURITY", 0L);
        issuesByType.put("PERFORMANCE", 0L);
        issuesByType.put("CODE_QUALITY", 0L);

        Map<String, Long> issuesBySeverity = new HashMap<>();
        issuesBySeverity.put("CRITICAL", 0L);
        issuesBySeverity.put("HIGH", 0L);
        issuesBySeverity.put("MEDIUM", 0L);
        issuesBySeverity.put("LOW", 0L);

        Map<String, Double> resolutionRate = new HashMap<>();
        List<DashboardController.TopIssue> topIssues = new ArrayList<>();
        Map<String, Long> issuesByFile = new HashMap<>();

        return new DashboardController.IssueStatistics(
                issuesByType,
                issuesBySeverity,
                resolutionRate,
                topIssues,
                issuesByFile
        );
    }

    /**
     * Obter métricas de performance
     */
    public DashboardController.PerformanceMetrics getPerformanceMetrics(int days) {
        log.debug("Gerando métricas de performance para {} dias", days);

        List<DashboardController.PerformancePoint> timeline = new ArrayList<>();

        return new DashboardController.PerformanceMetrics(
                2500.0, // TODO: Tempo médio de análise
                99.5, // TODO: Uptime do sistema
                1024L, // TODO: Uso de memória
                15.5, // TODO: Uso de CPU
                1000L, // TODO: Contagem de requests
                0.5, // TODO: Taxa de erro
                timeline
        );
    }

    /**
     * Obter alertas ativos
     */
    public List<DashboardController.SystemAlert> getActiveAlerts() {
        log.debug("Buscando alertas ativos");

        List<DashboardController.SystemAlert> alerts = new ArrayList<>();
        
        // TODO: Implementar sistema de alertas
        // Exemplo de alerta
        alerts.add(new DashboardController.SystemAlert(
                "alert_001",
                "PERFORMANCE",
                "WARNING",
                "Tempo de análise acima do normal",
                "Análises estão levando mais de 5 segundos em média",
                LocalDateTime.now().minusHours(2),
                false
        ));

        return alerts;
    }

    /**
     * Obter dados para gráficos
     */
    public DashboardController.ChartData getChartData(String chartType, int days, Long repositoryId) {
        log.debug("Gerando dados do gráfico: {} para {} dias", chartType, days);

        List<String> labels = new ArrayList<>();
        List<DashboardController.DataSeries> datasets = new ArrayList<>();
        Map<String, Object> options = new HashMap<>();

        switch (chartType.toLowerCase()) {
            case "reviews-timeline":
                labels.addAll(List.of("Jan", "Feb", "Mar", "Apr", "Mai", "Jun"));
                datasets.add(new DashboardController.DataSeries(
                        "Reviews",
                        List.of(10, 15, 12, 20, 18, 25),
                        "#4CAF50",
                        "#4CAF50",
                        Map.of("fill", false)
                ));
                break;
                
            case "quality-distribution":
                labels.addAll(List.of("Excelente", "Bom", "Regular", "Ruim"));
                datasets.add(new DashboardController.DataSeries(
                        "Distribuição",
                        List.of(25, 45, 20, 10),
                        "#2196F3",
                        "#2196F3",
                        Map.of("type", "pie")
                ));
                break;
                
            default:
                log.warn("Tipo de gráfico não reconhecido: {}", chartType);
        }

        return new DashboardController.ChartData(chartType, labels, datasets, options);
    }

    /**
     * Exportar dados do dashboard
     */
    public String exportDashboard(DashboardController.ExportRequest request) {
        log.info("Exportando dashboard no formato: {}", request.format());

        // TODO: Implementar exportação completa
        switch (request.format().toUpperCase()) {
            case "PDF":
                return generatePdfExport(request);
            case "CSV":
                return generateCsvExport(request);
            case "JSON":
                return generateJsonExport(request);
            default:
                throw new IllegalArgumentException("Formato de exportação não suportado: " + request.format());
        }
    }

    // Métodos auxiliares privados

    private List<DashboardController.QuickStat> generateQuickStats(LocalDateTime cutoffDate) {
        List<DashboardController.QuickStat> stats = new ArrayList<>();

        stats.add(new DashboardController.QuickStat(
                "Reviews Hoje",
                codeReviewRepository.countByCreatedAtAfter(LocalDateTime.now().withHour(0).withMinute(0)),
                "reviews",
                "up",
                12.5
        ));

        stats.add(new DashboardController.QuickStat(
                "Score Médio",
                codeReviewRepository.findAverageScoreAfter(cutoffDate),
                "pontos",
                "stable",
                0.0
        ));

        return stats;
    }

    private String generatePdfExport(DashboardController.ExportRequest request) {
        // TODO: Implementar geração de PDF
        return "base64-encoded-pdf-content";
    }

    private String generateCsvExport(DashboardController.ExportRequest request) {
        // TODO: Implementar geração de CSV
        return "csv,content,here";
    }

    private String generateJsonExport(DashboardController.ExportRequest request) {
        // TODO: Implementar geração de JSON
        return "{\"data\": \"json-content\"}";
    }
}
